dependency-analysis/main.go

```diff
@@ -0,0 +1,349 @@
+		checks.Repo.Name,
+		checks.Repo.Name))
+	sb.WriteString("<table>\n")
+	sb.WriteString("<tr><th>Check</th><th>Score</th></tr>\n")
+	for _, check := range checks.Checks {
+		sb.WriteString(fmt.Sprintf("<tr><td>%s</td><td>%d</td></tr>\n", check.Name, check.Score))
+	}
+	sb.WriteString("</table>\n")
+	if len(checks.Vulnerabilities) > 0 {
+		sb.WriteString("<table>\n")
+		sb.WriteString("<tr><th>Vulnerability</th><th>Severity</th><th>Summary</th></tr>\n")
+		for _, vulns := range checks.Vulnerabilities {
+			sb.WriteString(
+				fmt.Sprintf("<tr><td>%s</td><td>%s</td><td>%s</td></tr>\n", vulns.AdvisoryURL, vulns.Severity, vulns.AdvisorySummary)) //nolint:lll
+		}
+		sb.WriteString("</table>\n")
+	}
+
+	sb.WriteString("</details>")
+	return sb.String()
+}
+
+// GetDependencyDiff returns the dependency diff between two commits.
+// It returns an error if the dependency graph is not enabled.
+func GetDependencyDiff(owner, repo, token, base, head string) (*[]DependencyDiff, error) {
+	message := "failed to get dependency diff, please enable dependency graph https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-the-dependency-graph" //nolint:lll
+	if owner == "" {
+		return nil, fmt.Errorf("owner is required") //nolint:goerr113
+	}
+	if repo == "" {
+		return nil, fmt.Errorf("repo is required") //nolint:goerr113
+	}
+	if token == "" {
+		return nil, fmt.Errorf("token is required") //nolint:goerr113
+	}
+	resp, err := GetGitHubDependencyDiff(owner, repo, token, base, head)
+	defer resp.Body.Close() //nolint:staticcheck
+
+	if resp.StatusCode != http.StatusOK {
+		// if the dependency graph is not enabled, we can't get the dependency diff
+		return nil,
+			fmt.Errorf(" %s: %v", message, resp.Status) //nolint:goerr113
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dependency diff: %w", err)
+	}
+
+	var data []DependencyDiff
+	err = json.NewDecoder(resp.Body).Decode(&data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode dependency diff: %w", err)
+	}
+	// filter out the dependencies that are not added
+	var filteredData []DependencyDiff
+	for _, dep := range data { //nolint:gocritic
+		// also if the source repo doesn't start with GitHub.com, we can ignore it
+		if dep.ChangeType == "added" && dep.SourceRepositoryURL != "" &&
+			strings.HasPrefix(dep.SourceRepositoryURL, "https://github.com") {
+			filteredData = append(filteredData, dep)
+		}
+	}
+	return &filteredData, nil
+}
+
+// GetGitHubDependencyDiff returns the dependency diff between two commits.
+// It returns an error if the dependency graph is not enabled.
+func GetGitHubDependencyDiff(owner, repo, token, base, head string) (*http.Response, error) {
+	req, err := http.NewRequest("GET",
+		fmt.Sprintf("https://api.github.com/repos/%s/%s/dependency-graph/compare/%s...%s", owner, repo, base, head), nil)
```

Comment:
Can the GitHub client from `go-github` be used?

Reply:
The go-github doesn't provide this API, and that was the reason for me to use this.